src/config/include/config/Node.hpp

```diff
@@ -0,0 +1,444 @@
+                }
+                return Node{_parent->_yaml, _parent->_parent, _parent->_valid, childKey.str()};
+            }
+        }
+        auto yaml = this->yamlGet(key);
+        if (yaml) {
+            return Node{*yaml, this, true, keyStr};
+        } else {
+            return Node{YAML::Node{}, this, false, keyStr};
+        }
+    }
+
+    void appendKey(std::ostringstream& out) const;
+
+    YAML::Node _yaml;
+    std::string _key;
+    const Node* _parent;
+    bool _valid;
+};
+
+
+// we can act like a Node if iterated value is a scalar or
+// we can act like a pair of Nodes if iterated value is a map entry
+class IteratorValue : public std::pair<Node, Node>, public Node {
+    using NodePair = std::pair<Node, Node>;
+
+public:
+    // jump through immense hoops to avoid knowing anything about the actual yaml iterator other
+    // than its pair form is a pair of {YAML::Node, YAML::Node}
+    template <typename ITVal>
+    IteratorValue(const Node* parent, ITVal itVal, size_t index)
+        : NodePair{std::make_pair(
+              Node{itVal.first,
+                   parent,
+                   itVal.first,
+                   itVal.first ? (itVal.first.template as<std::string>() + "$key") : ""},
+              Node{itVal.second,
+                   parent,
+                   itVal.second,
+                   itVal.first ? itVal.first.template as<std::string>() : v1::toString(index)})},
+          Node{itVal, parent, itVal, v1::toString(index)} {}
+};
+
+
+class Node::iterator {
+    // Don't expose or assume the type of YAML::Node.begin()
+    using IterType = decltype(std::declval<const YAML::Node>().begin());
+
+    IterType _child;
+    const Node* parent;
+    size_t index = 0;
+
+    iterator(IterType child, const Node* parent) : _child(std::move(child)), parent(parent) {}
+
+    friend Node;
+
+public:
+    auto operator++() {
+        ++index;
+        return _child.operator++();
+    }
+
+    auto operator*() const {
+        auto out = _child.operator*();
+        return IteratorValue{parent, out, index};
+    }
+
+    auto operator-> () const {
+        auto out = _child.operator*();
+        return IteratorValue{parent, out, index};
+    }
+
+    auto operator==(const iterator& rhs) const {
+        return _child == rhs._child;
+    }
+
+    auto operator!=(const iterator& rhs) const {
+        return _child != rhs._child;
+    }
+};
+
+template <typename T, typename F>
+std::vector<T> Node::getPlural(const std::string& singular,
+                               const std::string& plural,
+                               F&& f) const {
+    std::vector<T> out;
+    auto pluralValue = (*this)[plural];
+    auto singValue = (*this)[singular];
+    if (pluralValue && singValue) {
+        std::stringstream str;
+        str << "Can't have both '" << singular << "' and '" << plural << "'. ";
+        str << "Path: '" << this->path() << "$plural(" << singular << "," << plural << ")'";
```

Comment:
Does "$plural' have a special meaning here?

Reply:
nah this is just to hopefully make it clear what the code is looking for:

```c++
    {
        Node node{"{Foo: 8, Foos: [1,2]}", ""};
        REQUIRE_THROWS_WITH([&]() { node.getPlural<int>("Foo", "Foos"); }(),
                            Catch::Matches("Can't have both 'Foo' and 'Foos'. Path: '\\$plural\\(Foo,Foos\\)'"));
    }
```